homotopy-core/src/graph.rs

```diff
@@ -1,310 +1,295 @@
+        graph
     }
 
-    for e in graph.edge_references() {
-        let rewrite: &RewriteN = (&e.weight().1).try_into()?;
-
-        let source_slices = &node_to_slices[e.source()];
-        let source_size = source_slices.len();
-        let target_slices = &node_to_slices[e.target()];
-        let target_size = target_slices.len();
-
-        // Identity rewrite between source slices
-        if let (Some(s), Some(t)) = (source_slices[0], target_slices[0]) {
-            if let Some(key) = edge_map(
-                Some(e.id()),
-                RewriteOrigin::Sparse,
-                Some(e.weight().0.clone()),
-            ) {
-                exploded_graph.add_edge(s, t, (key, Rewrite::identity(rewrite.dimension() - 1)));
+    /// Explodes a slice graph to obtain the slice graph one dimension higher.
+    pub fn explode<F, G, H, V2, E2, Ix2>(
+        &self,
+        mut node_map: F,
+        mut internal_edge_map: G,
+        mut external_edge_map: H,
+    ) -> Result<ExplosionOutput<V2, E2, Ix, Ix2>, DimensionError>
+    where
+        Ix2: IndexType,
+        F: FnMut(NodeIndex<Ix>, &V, SliceIndex) -> Option<V2>,
+        G: FnMut(NodeIndex<Ix>, &V, RewriteOrigin) -> Option<E2>,
+        H: FnMut(EdgeIndex<Ix>, &E, RewriteOrigin) -> Option<E2>,
+    {
+        let mut graph = SliceGraph::default();
+
+        let mut nodes = IdxVec::splat(vec![], self.node_count());
+        let mut internal_edges = IdxVec::splat(vec![], self.node_count());
+        let mut external_edges = IdxVec::splat(vec![], self.edge_count());
+
+        for (n, (key, diagram)) in self.node_references() {
+            let diagram: &DiagramN = diagram.try_into()?;
+
+            let mut add_node = |si: SliceIndex, slice: Diagram| {
+                nodes[n].push(|| -> Option<_> {
+                    let key = node_map(n, key, si)?;
+                    graph.add_node((key, slice)).into()
+                }());
+            };
+
+            // Source slice
+            add_node(Boundary::Source.into(), diagram.source());
+
+            // Interior slices
+            for (i, slice) in diagram.slices().enumerate() {
+                add_node(Height::from(i).into(), slice);
             }
-        }
 
-        // Identity rewrite between target slices
-        if let (Some(s), Some(t)) = (
-            source_slices[source_size - 1],
-            target_slices[target_size - 1],
-        ) {
-            if let Some(key) = edge_map(
-                Some(e.id()),
-                RewriteOrigin::Sparse,
-                Some(e.weight().0.clone()),
-            ) {
-                exploded_graph.add_edge(s, t, (key, Rewrite::identity(rewrite.dimension() - 1)));
+            // Target slice
+            add_node(Boundary::Target.into(), diagram.target());
+
+            let mut add_edge =
+                |si: SliceIndex, ti: SliceIndex, ro: RewriteOrigin, rewrite: Rewrite| {
+                    let si = (si.to_int(diagram.size()) + 1) as usize;
```

Comment:
You can't implement `From<SliceIndex>` for `{integer}` because `From<_>` is a foreign trait and `{integer`} is a foreign type. You could use an `Into<_>` instance though?

Reply:
We can't do that because to_int takes an argument.

Reply:
Ah - of course